packages/ember-glimmer/tests/integration/syntax/each-in-test.js

```diff
@@ -305,3 +305,29 @@ moduleFor('Syntax test: {{#each-in}} undefined path', class extends RenderingTes
     this.assertText('');
   }
 });
+
+moduleFor('Syntax test: {{#each-in}}  for array', class extends RenderingTest {
+  ['@test it iterate over array with `in` instead of walking over elements'](assert) {
+    let arr = [1, 2, 3];
+    arr.someKey = 'someKeyValue';
+
+    this.render(strip`
+      {{#each-in arr as |key value|}}
+        [{{key}}:{{value}}]
+      {{/each-in}}`, { arr: [] });
+
+    this.assertText('');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('');
+
+    this.runTask(() => set(this.context, 'arr', arr));
+
+    this.assertText('[0:1][1:2][2:3][someKey:someKeyValue]');
+
+    this.runTask(() => set(this.context, 'arr', []));
```

Comment:
For the "R" (reset with replacement) step, you want to revert the output to the initial state, but with a new/fresh object. So if you did what I suggested in the beginning and started with `arr`, you would want to make another instance of an array with the same objects/keys (don't reuse the instance you mutated in the "U" step).